### PR TITLE
TG-672 Refactor and small functional changes to core string refinement decision procedure

### PR DESCRIPTION
**Pull request by jasigal**
Each commit is one type of change for ease of reviewing.

### CODE REVIEW

**Comment by jasigal** [comment]
@romainbrenguier @LAJW @peterschrammel would you please review time permitting?

File: src/solvers/refinement/string_refinement.cpp [hunk 1]
```diff
@@ -96,38 +96,40 @@ static std::vector<exprt> instantiate_not_contains(
...
   mp_integer out;
   if(to_integer(expr, out))
     return { };
   return out;
 }
 
-template<>
-optionalt<std::size_t> expr_cast<std::size_t>(const exprt &expr)
+optionalt<std::size_t> expr_cast(
+  const exprt &expr, type_wrappert<std::size_t>)
```
Comment (LAJW) [inline src/solvers/refinement/string_refinement.cpp]: That change should be reverted. It is a workaround for a bug in CLion linter. [This template is specialized correctly](http://en.cppreference.com/w/cpp/language/template_specialization).
Should erroneous CLion warnings be taken into consideration, version based on class/struct templates should be tried first, before using this. Verbosity must not be added to client code.

File: src/solvers/refinement/string_refinement.cpp [hunk 2]
```diff
@@ -231,28 +229,6 @@ static void display_index_set(
          << " newly added)" << eom;
 }
 
-/// compute the index set for all formulas, instantiate the formulas with the
-/// found indexes, and add them as lemmas.
-
-static void display_current_index_set(
```
Comment (LAJW) [inline src/solvers/refinement/string_refinement.cpp:236]: I think @romainbrenguier should comment on whether this is in active use while debugging string refinement behaviour.

**Review by LAJW — CHANGES_REQUESTED** [review]

File: src/solvers/refinement/string_refinement.cpp (hunk 2 above)
Comment (jasigal) [reply src/solvers/refinement/string_refinement.cpp:236]: They were used right next to each other, and the `display_index_set` function highlights which indices are new.

File: src/solvers/refinement/string_refinement.cpp (hunk 1 above)
Comment (jasigal) [reply src/solvers/refinement/string_refinement.cpp]: Added your changes now.

**Review by LAJW — APPROVED** [review]

File: src/solvers/refinement/string_refinement.cpp (hunk 2 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_refinement.cpp:236]: Yes I agree `display_index_set` is more useful

File: src/solvers/refinement/string_refinement.cpp [hunk 3]
```diff
@@ -1855,11 +1666,39 @@ static void initial_index_set(
...
+      // cur is of the form s[i] and no quantified variable appears in i
+      add_to_index_set(index_set, ns, s, i);
+
+      it.next_sibling_or_parent();
+    }
+    else
+      ++it;
+  }
+
+  minus_exprt kminus1(
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_refinement.cpp]: `const` ?

File: src/solvers/refinement/string_refinement.cpp [hunk 4]
```diff
@@ -738,43 +582,50 @@ decision_proceduret::resultt string_refinementt::dec_solve()
...
-  update_index_set(index_set, current_index_set, ns, cur);
-  cur.clear();
-  for(const auto &lemma :
-        generate_instantiations(debug(), current_index_set, universal_axioms))
-    add_lemma(lemma);
-  display_current_index_set(debug(), ns, current_index_set);
+  initial_index_set(
+    index_sets,
+    ns,
+    axioms);
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_refinement.cpp]: this can fit in one line

**Review by romainbrenguier — APPROVED** [review]
Looks good

**Comment by jasigal** [comment]
Test-gen PR not passing, don't merge.

**Comment by jasigal** [comment]
Test-gen passes.

**Pull request merged**